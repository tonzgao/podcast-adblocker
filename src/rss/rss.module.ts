import { Module } from '@nestjs/common';
import { RssService } from './rss.service';

@Module({
  imports: [],
  providers: [RssService],
  exports: [RssService],
})
export class RssModule {}
