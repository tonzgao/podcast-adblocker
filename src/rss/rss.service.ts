import { Injectable } from '@nestjs/common';

import Parser from 'rss-parser';

@Injectable()
export class RssService {
  parser: Parser;

  constructor() {
    this.parser = new Parser();
  }

  public parseUrl(url: string) {
    return this.parser.parseURL(url);
  }
}
