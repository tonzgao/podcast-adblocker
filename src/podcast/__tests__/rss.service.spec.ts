import { Test, TestingModule } from '@nestjs/testing';
import { ConfigModule } from '@nestjs/config';
import * as Parser from 'rss-parser';

import { RssService } from '../rss.service';

import example from './example'

describe('RssService', () => {
  let rssService: RssService;

  beforeEach(async () => {
    const app: TestingModule = await Test.createTestingModule({
      imports: [ConfigModule.forRoot({isGlobal: true})],
      controllers: [],
      providers: [RssService],
    }).compile();

    rssService = app.get<RssService>(RssService);
  });

  describe('root', () => {
    it('should throw on non-rss feed', async () => {
      await expect(rssService.proxyUrl('https://example.com')).rejects.toThrow('Feed not recognized as RSS 1 or 2.')
    });

    it('should replace rss feed urls', async () => {
      const parser = new Parser();
      rssService.parser = {
        parseString: parser.parseString,
        parseURL: async () => Promise.resolve(parser.parseString(example)),
      }
      expect(await rssService.proxyUrl('https://example.com')).toMatchSnapshot();
    });
  });
});
