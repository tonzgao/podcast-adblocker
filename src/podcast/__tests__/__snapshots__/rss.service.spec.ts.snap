// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RssService root should replace rss feed urls 1`] = `
"<items>
    <title>Top 10 myths about caring for a zebra</title>
    <pubDate>Tue, 14 Mar 2017 12:00:00 GMT</pubDate>
    <enclosure>
        <url>http://localhost:3000/download/https://www.example.com/podcasts/dafnas-zebras/audio/toptenmyths.mp3</url>
        <type>audio/mpeg</type>
        <length>34216300</length>
    </enclosure>
    <content>Here are the top 10 misunderstandings about the care, feeding, and breeding of these lovable striped animals.</content>
    <contentSnippet>Here are the top 10 misunderstandings about the care, feeding, and breeding of these lovable striped animals.</contentSnippet>
    <guid>dzpodtop10</guid>
    <isoDate>2017-03-14T12:00:00.000Z</isoDate>
    <itunes>
        <duration>30:00</duration>
    </itunes>
</items>
<items>
    <title>Keeping those stripes neat and clean</title>
    <pubDate>Fri, 24 Feb 2017 12:00:00 GMT</pubDate>
    <enclosure>
        <url>http://localhost:3000/download/https://www.example.com/podcasts/dafnas-zebras/audio/cleanstripes.mp3</url>
        <type>audio/mpeg</type>
        <length>26004388</length>
    </enclosure>
    <content>Keeping your zebra clean is time consuming, but worth the effort.</content>
    <contentSnippet>Keeping your zebra clean is time consuming, but worth the effort.</contentSnippet>
    <guid>dzpodclean</guid>
    <isoDate>2017-02-24T12:00:00.000Z</isoDate>
    <itunes>
        <duration>22:48</duration>
    </itunes>
</items>
<title>Dafna's Zebra Podcast</title>
<description>A pet-owner's guide to the popular striped equine.</description>
<link>https://www.example.com/podcasts/dafnas-zebras/</link>
<language>en-us</language>
<itunes>
    <owner>
        <email>dafna@example.com</email>
    </owner>
    <image>https://www.example.com/podcasts/dafnas-zebras/img/dafna-zebra-pod-logo.jpg</image>
    <author>Dafna</author>
</itunes>"
`;
